asab/api/discovery.py

```diff
@@ -33,81 +54,170 @@ async def locate(self, instance_id: str = None, service_id: str = None) -> list:
+		It is a dict of identifier types as keys and dict as values. The second-layer dict has identifiers as keys an a set of tuples as a value. Each tuple contains host and port.
+
+		Example of the data structure:
+			{
+				"instance_id": {
+					"lmio-receiver-1": {("node1", 1234)},
+					"asab-remote-control-1": {("node2", 1234)}
+					...
+				},
+				"service_id": {
+					"lmio-receiver": {("node1", 1234), ("node2", 1234), ("node3", 1234)},
+					...
+				},
+				"custom1_id": {
+					"myid123": {("node1", 5678), ("node2", 5678)},
+					...
+				},
+				"custom2_id": {
+					"myid123": {("node1", 5678), ("node2", 5678)},
+					...
+				},
+				...
+			}
+		"""
+		if self.CacheLock.locked():
+			# Only one cache update at a time
+			return
+
+		async with self.CacheLock:
+
+			advertised = {
+				"instance_id": {},
+				"service_id": {},
+			}
+			async for item, item_data in self._iter_zk_items("/run"):
+				instance_id = item_data.get("instance_id")
+				service_id = item_data.get("service_id")
+				discovery: typing.Dict[str, list] = item_data.get("discovery", {})
+
+				if instance_id is not None:
+					discovery["instance_id"] = [instance_id]
+
+				if instance_id is not None:
+					discovery["service_id"] = [service_id]
+
+				web = item_data.get("web")
+				host = item_data.get("node_id", item_data.get("host"))
+
+				if web is None or host is None:
+					continue
+				for i in web:
+					try:
+						ip = i[0]
+						port = i[1]
+					except KeyError:
+						L.error("Unexpected format of 'web' section in advertised data: '{}'".format(web))
+						return
+					if ip not in ("0.0.0.0", "::"):
+						continue
+
+					if discovery is not None:
+						for id_type, ids in discovery.items():
+							if advertised.get(id_type) is None:
+								advertised[id_type] = {}
+							for identifier in ids:
+								if identifier is not None:
+									if advertised[id_type].get(identifier) is None:
+										advertised[id_type][identifier] = {(host, port)}
+									else:
+										advertised[id_type][identifier].add((host, port))
+
+			if self.AdvertisedCache is None:
+				# When application starts, Auth Service needs to load public keys. Cache is not ready during `initialize` to discover auth server.
+				# This signal makes auth service to repeat the call to auth server when discovery session is ready and load public keys.
+				self.AdvertisedCache = advertised
+				self.DiscoveryReady.set()
+			else:
+				self.AdvertisedCache = advertised
+
 
 	async def _iter_zk_items(self, path):
 		base_path = self.ZooKeeperContainer.Path + path
-		items = await self.ZooKeeperContainer.ZooKeeper.get_children(base_path)
+
+		def get_items():
+			try:
+				return self.ZooKeeperContainer.ZooKeeper.Client.get_children(base_path, watch=self._update_cache)
+			except (kazoo.exceptions.SessionExpiredError, kazoo.exceptions.ConnectionLoss):
+				# TODO: this is silent error
```

Comment:
Don't do silent errors - print warnings.

Reply:
silent in receiver